<h3>BREAST CANCER PREDICTION USING LOGISTIC REGRESSION<h3>

<p1>The objective of this project is to analyze and classify breast tumors as malignant or benign using machine learning techniques, specifically logistic regression. The Breast Cancer Wisconsin (Diagnostic) Dataset is used for this purpose.<p1>

<p1>Breast cancer is a prevalent form of cancer among women, accounting for a significant percentage of all cancer cases worldwide. Detecting and classifying tumors as cancerous (malignant) or non-cancerous (benign) is crucial for accurate diagnosis and treatment.<p1>

<p1>The dataset provides features related to breast tumor characteristics, such as texture, perimeter, area, smoothness, and more. The challenge is to develop classification models that can accurately predict the type of cancer based on these features.<p1>

<p1>We will try to understand, explore and cleanup the dataset. Next we will build a logistic regression model, a statistical technique used for binary classification. We will train the model on the dataset, using the features to predict whether a tumor is malignant or benign.

Hyperparameters of the model will also be tuned, to optimize its performance. We will adjust parameters such as the regularization strength, solver algorithm, and maximum number of iterations to enhance the model's accuracy.

The performance of the logistic regression model will be done using evaluation metrics such as accuracy, precision, recall, and F1-score. These metrics provide insights into the model's ability to correctly classify malignant and benign tumors.<p1>

In [64]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import classification_report, roc_auc_score

In [65]:
#read the data
data = pd.read_csv('/content/breast-cancer.csv')

In [66]:
#check the first three rows of the data set
data.head(3)

id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38           122.8     1001.0   
1    842517         M        20.57         17.77           132.9     1326.0   
2  84300903         M        19.69         21.25           130.0     1203.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33            184.6      2019.0   
1  ...         24.99          23.41            158.8      1956.0   
2  ...         23.57          25.53            152.5      1709.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  

[3 rows x 32 columns]

In [67]:
# check some statistical description of the data
data.describe()

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

<p1>We will create an extra column to have our class 0 and 1. 0 to represent malignant cases and 1 to represent benign cases<p1>

In [68]:
data['label'] = data['diagnosis'].apply(lambda x: 1 if x == 'B' else 0)

In [69]:
# check data information 
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 569 entries, 0 to 568
Data columns (total 33 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   id                       569 non-null    int64  
 1   diagnosis                569 non-null    object 
 2   radius_mean              569 non-null    float64
 3   texture_mean             569 non-null    float64
 4   perimeter_mean           569 non-null    float64
 5   area_mean                569 non-null    float64
 6   smoothness_mean          569 non-null    float64
 7   compactness_mean         569 non-null    float64
 8   concavity_mean           569 non-null    float64
 9   concave points_mean      569 non-null    float64
 10  symmetry_mean            569 non-null    float64
 11  fractal_dimension_mean   569 non-null    float64
 12  radius_se                569 non-null    float64
 13  texture_se               569 non-null    float64
 14  perimeter_se             5

<h3>Exploratory Data Analysis<h3>
<p1>We try to ascertain the percentage of malignant cases to benign cases.<p1>

In [71]:
#count the values of bening and malignant
data['label'].value_counts()

1    357
0    212
Name: label, dtype: int64

In [72]:
# get the percentage of malignant and bening cases
percentage_1s = (data['label'] == 1).mean() * 100
percentage_0s = (data['label'] == 0).mean() * 100

# Print the percentages
print(f"Percentage of 1s: {percentage_1s:.2f}%")
print(f"Percentage of 0s: {percentage_0s:.2f}%")

Percentage of 1s: 62.74%
Percentage of 0s: 37.26%


next lets group the features according to the labels, and take the mean. We will further drop the ID column since its a unique ID that identifies the patients or samples and not necessarily an important feature.

In [73]:
# drop column ID
data = data.drop('id', axis=1)
data = data.drop('diagnosis', axis=1)

In [74]:
# grouping the features.
data.groupby('label').mean()

radius_mean  texture_mean  perimeter_mean   area_mean  smoothness_mean  \
label                                                                           
0        17.462830     21.604906      115.365377  978.376415         0.102898   
1        12.146524     17.914762       78.075406  462.790196         0.092478   

       compactness_mean  concavity_mean  concave points_mean  symmetry_mean  \
label                                                                         
0              0.145188        0.160775             0.087990       0.192909   
1              0.080085        0.046058             0.025717       0.174186   

       fractal_dimension_mean  ...  radius_worst  texture_worst  \
label                          ...                                
0                    0.062680  ...     21.134811      29.318208   
1                    0.062867  ...     13.379801      23.515070   

       perimeter_worst   area_worst  smoothness_worst  compactness_worst  \
label                                                                      
0           141.370330  1422.286321          0.144845           0.374824   
1            87.005938   558.899440          0.124959           0.182673   

       concavity_worst  concave points_worst  symmetry_worst  \
label                                                          
0             0.450606              0.182237        0.323468   
1             0.166238              0.074444        0.270246   

       fractal_dimension_worst  
label                           
0                     0.091530  
1                     0.079442  

[2 rows x 30 columns]

<h3>Creating Input Features, Split and Train Dataset<h3>

<p1>On visual inspection we can see that in certain features, the values in the malignant cases are somewhat higher than in benign cases.
Next we set our input features and target features. The input features are our x variables and the target variables is our y variable. We also drop the diagnosis column because it is a category variable and will raise error when being computed on the model.<p1>

In [77]:
# creating input and target features
X = data.drop('label', axis=1)
y = data['label']

In [78]:
# create test and training set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

Next we train our model. In the constructor for the logistic regression model, we will use the solver liblinear to avoid non convergence problems. It can be a good choice, especially when dealing with binary classification problems.

The 'liblinear' solver is specifically designed for logistic regression and can handle both L1 and L2 regularization. It works well for small to medium-sized datasets and is generally a reliable solver choice for binary classification tasks.

In [79]:
# Create an instance of the logistic regression model
logreg = LogisticRegression(solver='liblinear', max_iter=25)


# Fit the model to the training data
logreg.fit(X_train, y_train)

# Predict labels for the test data
y_pred = logreg.predict(X_test)

# Calculate accuracy on the test data
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.956140350877193


The accuracy score indicates the proportion of correctly predicted labels compared to the total number of samples in the test data. By comparing the predicted labels (y_pred) with the true labels (y_test), the accuracy score is helps you assess how well the model is performing on the test data.

It's important to note that accuracy alone might not provide the complete picture of a model's performance, especially in imbalanced datasets. It's often useful to consider other evaluation metrics such as precision, recall, or F1 score to gain a more comprehensive understanding of the model's behavior.

Overall, obtaining a high accuracy score indicates that the logistic regression model is performing well on the test data.

Next we will build a predictive system where we can supply input X and allow the system predict a class or label, that is if the tumour is malignant or benign.

In [80]:
#create an input array
input_data = (13.54,14.36,87.46,566.3,0.09779,0.08129,0.06664,0.04781,0.1885,0.05766,0.2699,0.7886,2.058,23.56,0.008462,0.0146,0.02387,0.01315,0.0198,0.0023,15.11,19.26,99.7,711.2,0.144,0.1773,0.239,0.1288,0.2977,0.07259)
input_data_as_np = np.asarray(input_data)
input_data_reshaped = input_data_as_np.reshape(1,-1)

Reshaping an array does not simply mean adding a column to a vector. It involves changing the shape or dimensions of the array while maintaining the total number of elements.

When reshaping an array, the shape of the array is modified, which can involve changing the number of rows and columns. The reshaping operation reorganizes the elements of the array in a new layout.

In the specific case of reshaping a 1-dimensional array into a 2-dimensional array with a single row (1 row and multiple columns), the reshaping process essentially adds an additional dimension to the array.

The reshaping step input_data_reshaped = input_data_as_np.reshape(1, -1) is performed to transform the 1-dimensional array into a 2-dimensional array with a single row and an automatically calculated number of columns. The reshape() function is used, where the first argument 1 indicates the desired number of rows (in this case, 1), and the second argument -1 indicates that the number of columns should be automatically determined based on the size of the original array.

This reshaping step is necessary because scikit-learn's predict() method expects the input data to be in a 2-dimensional format, where each row represents an individual sample and each column represents a feature. Reshaping the input data ensures that it is compatible with the model for making predictions.

After reshaping, input_data_reshaped will be a 2-dimensional NumPy array with a shape of (1, num_features), where num_features is the number of features in the original input data.

We can then pass input_data_reshaped to the predict() method of your trained logistic regression model to obtain predictions for this new data point.

In [84]:
# predicting new unseen data
predictions = logreg.predict(input_data_reshaped)
print(predictions)


[1]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


Our model accurately predicts the class as Benign. We carry out further comprehensive tests to be sure our model is performing well.

In [82]:
# Calculate precision, recall, and F1 score
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.97      0.91      0.94        43
           1       0.95      0.99      0.97        71

    accuracy                           0.96       114
   macro avg       0.96      0.95      0.95       114
weighted avg       0.96      0.96      0.96       114



In [83]:
# Calculate ROC AUC score
auc_score = roc_auc_score(y_test, y_pred)
print("ROC AUC Score:", auc_score)

ROC AUC Score: 0.9464461185718965


Precision: For the "0" class, the model has a precision of 0.97, which means that 97% of the instances predicted as "0" are actually "0". For the "1" class, the precision is 0.95, indicating that 95% of the instances predicted as "1" are actually "1". Higher precision values indicate fewer false positives.

Recall: The "0" class has a recall of 0.91, meaning that the model correctly identifies 91% of the actual "0" instances. The "1" class has a recall of 0.99, indicating that the model captures 99% of the actual "1" instances. Higher recall values indicate fewer false negatives.

F1-Score: The F1-score considers both precision and recall. For the "0" class, the F1-score is 0.94, and for the "1" class, it is 0.97. The weighted average F1-score is 0.96, indicating a good overall balance between precision and recall.

Accuracy: The overall accuracy of the model is 0.96, which means that it correctly predicts the class label for 96% of the instances in the test data.

These metrics suggest that the logistic regression model is performing well in classifying both the "0" and "1" classes, with high precision, recall, and F1-score. The accuracy score confirms the model's overall effectiveness in predicting the class labels accurately.

The ROC AUC (Receiver Operating Characteristic Area Under the Curve) score is a metric that evaluates the performance of a binary classification model. It measures the ability of the model to distinguish between the positive and negative classes by analyzing the trade-off between the true positive rate (sensitivity) and the false positive rate (1 - specificity).

In this case, the ROC AUC score is 0.946, which indicates that the logistic regression model has a good discriminatory power. A score of 0.5 represents a random guess, while a score of 1.0 represents a perfect classifier.

A ROC AUC score of 0.946 suggests that the model performs well in distinguishing between the "0" and "1" classes, with a high true positive rate and a low false positive rate. This indicates that the model has a high probability of assigning higher predicted probabilities to the positive class compared to the negative class.

Overall, a ROC AUC score of 0.946 demonstrates that the logistic regression model is effective in classifying the instances and has a strong ability to discriminate between the two classes.

<h3>Conclusion<h3>

<p1>In conclusion, this project focused on utilizing logistic regression for the binary classification task of predicting breast tumor types as malignant or benign. By analyzing the Breast Cancer Wisconsin (Diagnostic) Dataset, we aimed to develop a model capable of accurately classifying tumors based on their characteristics.

Through data exploration, preprocessing, and model training, we built a logistic regression model and fine-tuned its hyperparameters to optimize performance. The evaluation metrics, including accuracy, precision, recall, and F1-score, were used to assess the model's effectiveness in correctly classifying tumors.

Overall, the logistic regression model demonstrated promising results, achieving a high level of accuracy and exhibiting good performance in different evaluation metrics. This suggests that logistic regression can be a valuable tool in aiding the diagnosis of breast cancer by classifying tumors as malignant or benign.

However, it is important to note that further analysis, validation, and refinement may be required before deploying the model in a real-world clinical setting. Additionally, exploring other classification algorithms, such as Support Vector Machines (SVMs), could provide valuable insights and potential improvements to the classification task.

In summary, this project contributes to the ongoing efforts in leveraging machine learning techniques for breast cancer diagnosis and emphasizes the potential of logistic regression as a reliable approach in this domain.<p1>